R/read-eval.R

```diff
@@ -0,0 +1,176 @@
+#' Parse R code.
+#'
+#' These functions parse_expr and transform text into R expressions.
+#' This is the first step of interpreting or evaluating a piece of
+#' code written by a programmer.
+#'
+#' \code{parse_expr()} returns one expression. If the text contains
+#' more than one expression (separated by colons or new lines), an
+#' error is issued. On the other hand \code{parse_exprs()} can handle
+#' multiple expressions. It always returns a list of expressions
+#' (compare to \code{\link[base]{parse}()} which returns an obsolete
+#' expression vector). All functions also support R connections (and
+#' will close the connection after parsing).
+#'
+#' The versions prefixed with \code{f_} return expressions quoted in
+#' formulas rather than raw expressions.
+#'
+#' @param x Text containing expressions to parse_expr for
+#'   \code{parse_expr()} and \code{parse_exprs()}. Can also be an R
+#'   connection, for instance to a file. Note that the connection will
+#'   be closed as side effect.
+#' @param env The environment for the formulas. Defaults to the
+#'   context in which the parse_expr function was called. Can be any object
+#'   with a \code{as_env()} method.
+#' @return \code{parse_expr()} returns a formula, \code{parse_exprs()}
+#'   returns a list of formulas.
+#' @seealso \code{\link[base]{parse}()}
+#' @export
+#' @examples
+#' # parse_expr() can parse_expr any R expression:
+#' parse_expr("mtcars %>% dplyr::mutate(cyl_prime = cyl / sd(cyl))")
+#'
+#' # A string can contain several expressions separated by ; or \n
+#' parse_exprs("NULL; list()\n foo(bar)")
+#'
+#' # The versions prefixed with f_ return formulas:
+#' f_parse_expr("foo %>% bar()")
+#' f_parse_exprs("1; 2; mtcars")
+#'
+#' # The env argument is passed to as_env(). It can be e.g. a string
+#' # representing a scoped package environment:
+#' f_parse_expr("identity(letters)", env = env_empty())
+#' f_parse_exprs("identity(letters); mtcars", env = "base")
+#'
+#'
+#' # You can also parse source files by passing a R connection. Let's
+#' # create a file containing R code:
+#' path <- tempfile("my-file.R")
+#' cat("1; 2; mtcars", file = path)
+#'
+#' # This file can be parsed by first opening a connection.
+#' # Note that the connection is automatically closed after usage.
+#' parse_exprs(file(path))
+parse_expr <- function(x) {
+  exprs <- parse_exprs(x)
+
+  n <- length(exprs)
+  if (n == 0) {
+    stop("No expression to parse_expr", call. = FALSE)
+  } else if (n > 1) {
+    stop("More than one expression parsed_expr", call. = FALSE)
+  }
+
+  exprs[[1]]
+}
+#' @rdname parse_expr
+#' @export
+parse_exprs <- function(x) {
+  if (inherits(x, "connection")) {
+    exprs <- parse(file = x)
+    close(x)
```

Comment:
I have added a `temporary()` function that add a `tmp = TRUE` attribute to its argument. Then `parse_exprs()` uses `maybe_close()` to close connections. It only closes connections with a temporary attribute.

It is used like this:

```r
parse_exprs(temporary(file(path)))
```

`temporary()` could be useful in other contexts. Is this a good strategy? Should `temporary()` have another name? Maybe `transient()`? `deathbed()` if we want to get dramatic :)